### PR TITLE
STORM-1705: Cap number of retries for a failed message

### PR DESCRIPTION
**Pull request by abhishekagarwal87**
This PR also makes retry strategy pluggable.

### CODE REVIEW

**Comment by knusbaum** [comment]
+1 LGTM

File: external/storm-kafka/pom.xml [hunk 1]
```diff
@@ -129,7 +129,7 @@
             <groupId>org.apache.storm</groupId>
             <artifactId>storm-core</artifactId>
             <version>${project.version}</version>
-            <scope>provided</scope>
+            <!--scope>provided</scope-->
```
Comment (HeartSaVioR) [inline external/storm-kafka/pom.xml]: Is it intended?

Comment (abhishekagarwal87) [inline external/storm-kafka/pom.xml]: good catch @HeartSaVioR . I had to make this change for in-IDE testing. I will revert this.

Comment (abhishekagarwal87) [inline external/storm-kafka/pom.xml]: Reverted the changes in pom.

File: external/storm-kafka/src/jvm/org/apache/storm/kafka/PartitionManager.java [hunk 2]
```diff
@@ -46,6 +46,9 @@
     private final CountMetric _fetchAPIMessageCount;
```
Comment (harshach) [inline external/storm-kafka/src/jvm/org/apache/storm/kafka/PartitionManager.java]: since we are refactoring this code. Can we simplify partitionManager constructor instead of the current state we can have few private methods for all the code in the constructor.

Comment (abhishekagarwal87) [inline external/storm-kafka/src/jvm/org/apache/storm/kafka/PartitionManager.java]: Agree that constructor is somewhat bloated. The whole class can use some refactoring. can we do it in a separate PR?

Comment (abhishekagarwal87) [inline external/storm-kafka/src/jvm/org/apache/storm/kafka/PartitionManager.java]: @harshach - is it fine with you to do the refactoring separately? Apologies for a long pause.

**Comment by abhishekagarwal87** [comment]
@harshach - ping

**Comment by harshach** [comment]
+1

**Comment by HeartSaVioR** [comment]
+1
Though it modifies public interface FailedMsgRetryManager, but it was only used to storm-kafka internally.
Btw, when we push this and release, we should respect backward compatibility for FailedMsgRetryManager since it will be used to pluggable implementation.

**Comment by abhishekagarwal87** [comment]
I agree @HeartSaVioR. Happy to incorporate feedback on the interface improve it further. One limitation I see is that interface takes offsets. It is not possible to add an implementation which can store failed messages in memory and return them (like new Kafka spout). But the interface still provides the control over actions on message retry e.g. one can log the offset in a log file if it exhausts the number of retries etc.

**Comment by abhishekagarwal87** [comment]
can this be merged? Please let me know if I need to raise a separate PR for master.

**Comment by HeartSaVioR** [comment]
@abhishekagarwal87 
Could you please add pluggable FailedMsgRetryManager to README of storm-kafka to let users know how to use? Thanks in advance.

File: external/storm-kafka/src/jvm/org/apache/storm/kafka/SpoutConfig.java [hunk 3]
```diff
@@ -37,6 +37,8 @@
     public long retryInitialDelayMs = 0;
     public double retryDelayMultiplier = 1.0;
     public long retryDelayMaxMs = 60 * 1000;
+    public int retryLimit = Integer.MAX_VALUE;
```
Comment (ptgoetz) [inline external/storm-kafka/src/jvm/org/apache/storm/kafka/SpoutConfig.java]: This is probably an unlikely edge case, but theoretically possible: If a tuple is retried `Integer.MAX_VALUE` times, it will be dropped.

To truly turn off retry limits, I would suggest setting this to `-1` and adding logic to check for that in logic that checks to see if the retry limit has been reached.

File: external/storm-kafka/src/jvm/org/apache/storm/kafka/ExponentialBackoffMsgRetryManager.java [hunk 4]
```diff
@@ -86,15 +94,23 @@ public Long nextFailedMessageToRetry() {
     }
 
     @Override
-    public boolean shouldRetryMsg(Long offset) {
+    public boolean shouldReEmitMsg(Long offset) {
         MessageRetryRecord record = this.records.get(offset);
         return record != null &&
                 this.waiting.contains(record) &&
                 System.currentTimeMillis() >= record.retryTimeUTC;
```
Comment (ptgoetz) [inline external/storm-kafka/src/jvm/org/apache/storm/kafka/ExponentialBackoffMsgRetryManager.java]: I would be nice to log a WARN message when we are dropping messages so users have a way of knowing when/why it is happening.

**Comment by ptgoetz** [comment]
Two minor nits. I'm +1 once they are addressed.

**Comment by abhishekagarwal87** [comment]
Thanks @ptgoetz and @HeartSaVioR . One thing I missed in my implementation is how to pass any additional configuration to failedMsgRetryManager. I can either add an extra HashMap or pass stormConf in the prepare method of failedMsgRetryManager. It's minor but still any suggestions?

**Comment by abhishekagarwal87** [comment]
I have taken care of comments and added documentation in storm-kafka readme. Please review the documentation as well.

**Comment by ptgoetz** [comment]
+1

**Comment by abhishekagarwal87** [comment]
@ptgoetz - unfortunately there was a bug in this PR. I have raised #1481 to fix that bug. Sorry about that.
